core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -155,50 +165,62 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+    * @param value amount of data in bytes or request processing time as a percentage
+    * @return throttle time in milliseconds
+    */
+  def maybeRecord(request: RequestChannel.Request, value: Double): Int = {
+    maybeRecord(request.session.sanitizedUser, request.header.clientId, value)
   }
 
-  def recordAndThrottleOnQuotaViolation(clientSensors: ClientSensors, value: Double, callback: Int => Unit): Int = {
+  def maybeRecord(sanitizedUser: String, clientId: String, value: Double): Int = {
     var throttleTimeMs = 0
-    try {
-      clientSensors.quotaSensor.record(value)
-      // trigger the callback immediately if quota is not violated
-      callback(0)
-    } catch {
-      case _: QuotaViolationException =>
-        // Compute the delay
-        val clientQuotaEntity = clientSensors.quotaEntity
-        val clientMetric = metrics.metrics().get(clientRateMetricName(clientQuotaEntity.sanitizedUser, clientQuotaEntity.clientId))
-        throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quota)).toInt
-        clientSensors.throttleTimeSensor.record(throttleTimeMs)
-        // If delayed, add the element to the delayQueue
-        delayQueue.add(new ThrottledResponse(time, throttleTimeMs, callback))
-        delayQueueSensor.record()
-        debug("Quota violated for sensor (%s). Delay time: (%d)".format(clientSensors.quotaSensor.name(), throttleTimeMs))
+
+    // Record metrics only if quotas are enabled.
+    if (quotasEnabled) {
+      val clientSensors = getOrCreateQuotaSensors(sanitizedUser, clientId)
+      try {
+        clientSensors.quotaSensor.record(value)
+      } catch {
+        case _: QuotaViolationException =>
+          // Compute the delay
+          val clientQuotaEntity = clientSensors.quotaEntity
+          val clientMetric = metrics.metrics().get(clientRateMetricName(clientQuotaEntity.sanitizedUser, clientQuotaEntity.clientId))
+          throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quota)).toInt
+          error("Quota violated for sensor (%s). Delay time: (%d)".format(clientSensors.quotaSensor.name(), throttleTimeMs))
+      }
     }
     throttleTimeMs
   }
 
+  /**
+    * Throttle a client by muting the associated channel for the given throttle time.
+    * @param request client request
+    * @param throttleTimeMs Duration in milliseconds for which the channel is to be muted.
+    * @return ThrottledChannel object
+    */
+  def maybeThrottle(request: RequestChannel.Request, throttleTimeMs: Int): Option[ThrottledChannel] = {
+    maybeThrottle(request.session.sanitizedUser, request.header.clientId, throttleTimeMs, request.channelUnmuteCallback)
+  }
+
+  def maybeThrottle(sanitizedUser: String, clientId: String, throttleTimeMs: Int, channelUnmuteCallback: () => Unit)
+    : Option[ThrottledChannel] = {
+    trace("maybeThrottle " +  sanitizedUser + " " + clientId + " " + throttleTimeMs)
+    if (quotasEnabled && throttleTimeMs > 0) {
```

Comment:
It seems a bit redundant to check `quotasEnabled` in both `maybeRecord` and `maybeThrottle`. It is probably more intuitive to check `quotasEnabled` only in `maybeRecord` such that `throttleTimeMs` will be 0 if `quotasEnabled` is false. Then `maybeThrottle` can act solely based on `throttleTimeMs`.

Reply:
Makes sense assuming that maybeThrottle() always uses throttle time returned by maybeRecord(), which is the case.